apf: request handler must wait for watch init goroutine to return

tkashem opened the pull request:
#### What type of PR is this?

/kind bug
/kind flake

#### What this PR does / why we need it:
request handler (in case it panics) must wait for watch initialization go routine to return

https://github.com/kubernetes/kubernetes/pull/126574 is causing the above bug to manifest and then is causing `TestApfWatchPanic` to flake

```
{Failed;  === RUN   TestApfWatchPanic
    priority-and-fairness_test.go:187: the APF handler has finished, checking atomicReadOnlyExecuting
    priority-and-fairness_test.go:189: Wanted 0 requests executing, got 1
--- FAIL: TestApfWatchPanic (0.00s)
;}
```

The flake in the test is due to:
https://github.com/kubernetes/kubernetes/blob/4cfdad0935cfb796d4219dbabe8eba669614df39/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L271-L279

If the handler (line 271) panics then it skips waiting for `<-resultCh`, this results in two go routines to race for `atomicReadOnlyExecuting`:
a) write (decrement  `atomicReadOnlyExecuting`):  https://github.com/kubernetes/kubernetes/blob/4cfdad0935cfb796d4219dbabe8eba669614df39/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L207

b) read (the test tries to read atomicReadOnlyExecuting): 
https://github.com/kubernetes/kubernetes/blob/4cfdad0935cfb796d4219dbabe8eba669614df39/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go#L188

`atomicReadOnlyExecuting` is protected by lock, but since there are no happens before relation the test flakes.


#### Which issue(s) this PR fixes:
Fixes #127048

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

tkashem commented:
I am going to share the stress test result

tkashem commented:
/assign @MikeSpreitzer

tkashem commented:
stress results
```
stress -p=256 ./filters.test -test.run=TestApfWatchPanic
5s: 281 runs so far, 0 failures
10s: 601 runs so far, 0 failures
15s: 1078 runs so far, 0 failures
...
4m5s: 20646 runs so far, 0 failures
4m10s: 21117 runs so far, 0 failures
4m15s: 21508 runs so far, 0 failures
```

```
stress -p=256 ./filters.test -test.run=TestApfWatchHandlePanic
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 321 runs so far, 0 failures
...
5m35s: 13003 runs so far, 0 failures
5m40s: 13236 runs so far, 0 failures
5m45s: 13464 runs so far, 0 failures
```

tkashem commented:
`stress` test results cc @MikeSpreitzer @liggitt 
```
stress -p=256 ./filters.test -test.run=TestApfWatchPanic
5s: 281 runs so far, 0 failures
10s: 601 runs so far, 0 failures
15s: 1078 runs so far, 0 failures
...
4m5s: 20646 runs so far, 0 failures
4m10s: 21117 runs so far, 0 failures
4m15s: 21508 runs so far, 0 failures
```

```
stress -p=256 ./filters.test -test.run=TestApfWatchHandlePanic
5s: 0 runs so far, 0 failures
10s: 256 runs so far, 0 failures
15s: 321 runs so far, 0 failures
...
5m35s: 13003 runs so far, 0 failures
5m40s: 13236 runs so far, 0 failures
5m45s: 13464 runs so far, 0 failures
```

tkashem commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:
```diff
@@ -266,17 +266,22 @@ func (h *priorityAndFairnessHandler) Handle(w http.ResponseWriter, r *http.Reque
-				panic(err)
-			}
+			func() {
+				defer func() {
```
self: although rare, we ca have a scenario where both the handler and the watch initialization goroutine panic, in case we have two panics we should combine both or let one take precedence, thoughts?

tkashem replied:
I have a separate PR https://github.com/kubernetes/kubernetes/pull/127075 in the works which wraps the panic from a spawned goroutine

aojea replied:
who panics here ? is `h.handler.ServeHTTP(w, watchReq)`?

tkashem replied:
yes, line 271, if `h.handler.ServeHTTP(w, watchReq)` panics then the handler skips the wait (line 277)

neolit123 replied:
> self: although rare, we can have a scenario where both the handler and the watch initialization goroutine panic, in case we have two panics we should combine both or let one take precedence, thoughts?

if the init goroutine panic is unexpected then it should be fixed instead of masking it?

neolit123 replied:
is there no other way to handle the panic here instead of the closure function and the defer?
also, why is `h.handler.ServeHTTP(w, watchReq)` panicing exactly?

i think the idea here is:
- an error happens
- the code wants to panic about the error so that the user is informed and the process is terminated

`h.handler.ServeHTTP(w, watchReq)` panicing seems out of place.

MikeSpreitzer replied:
(A) I think that the change here makes sense. It is protecting the integrity of the APF bookkeeping in the case where `h.handler` panics, while passing that panic along up the stack. Indeed, the panic should also be debugged and fixed. But until then, we still want APF to keep accurate books.

(B) I do not see how/why this fixes the observed problem with https://github.com/kubernetes/kubernetes/blob/5baa0623f9bc21ffcf04995ea6ff3f09e0ee49b8/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go#L556-L559 (I referenced the commit that `master` currently points to). In this test the panic comes up through https://github.com/kubernetes/kubernetes/blob/5baa0623f9bc21ffcf04995ea6ff3f09e0ee49b8/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L264 , not https://github.com/kubernetes/kubernetes/blob/5baa0623f9bc21ffcf04995ea6ff3f09e0ee49b8/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go#L271

dims commented:
cc @liggitt

tkashem commented:
I also ran `stress` on the entire package
```
$ stress ./filters.test -test.run=
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 0 runs so far, 0 failures
20s: 0 runs so far, 0 failures
25s: 0 runs so far, 0 failures
30s: 0 runs so far, 0 failures
35s: 0 runs so far, 0 failures
40s: 0 runs so far, 0 failures
45s: 0 runs so far, 0 failures
50s: 3 runs so far, 0 failures
...
3m20s: 32 runs so far, 0 failures
3m25s: 32 runs so far, 0 failures
3m30s: 34 runs so far, 0 failures
```

tkashem commented:
/cc @wojtek-t

aojea commented:
I think we should be cautious and avoid fix forward https://github.com/kubernetes/kubernetes/pull/127089

Jefftree commented:
/triage accepted

tkashem committed:
apiserver: fix apf watch test

the assert to verify that 'atomicReadOnlyExecuting' is zero
should be executed if the apf handler panics, all apf
bookkeeping must be completed before the handler returns

tkashem committed:
apiserver: all bookkeeping must complete before apf handler returns

all bookkeeping must complete before the apf handler returns,
whether it panics or returns normally

tkashem commented:
@MikeSpreitzer if I run with the changes in the test only https://github.com/kubernetes/kubernetes/pull/127029/commits/c2a249be40cbfd42004f47577753a808490424b8, the test fails and a data race exhibits
```
$ go test -race -v -count=1 k8s.io/apiserver/pkg/server/filters -run=TestApfWatchPanic
=== RUN   TestApfWatchPanic
==================
WARNING: DATA RACE
Write at 0x00000475e7f4 by goroutine 118:
  sync/atomic.AddInt32()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/race_amd64.s:281 +0xb
  sync/atomic.AddInt32()
      <autogenerated>:1 +0x14
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle.func7.deferwrap1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:207 +0x38
  runtime.deferreturn()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:602 +0x5d
  k8s.io/apiserver/pkg/server/filters.(*fakeWatchApfFilter).Handle()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:401 +0x127
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle.func8()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:264 +0x1e3

Previous read at 0x00000475e7f4 by goroutine 114:
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2.1.1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:188 +0x30
  runtime.gopanic()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/runtime/panic.go:770 +0x131
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:173 +0x2e
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:271 +0x11f2
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle-fm()
      <autogenerated>:1 +0x51
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2.1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:192 +0xc1
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:194 +0x312
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.WithRequestInfo.func4()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x17a
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.TestApfWatchPanic.func3()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:539 +0xc4
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  net/http.serverHandler.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3137 +0x2a1
  net/http.(*conn).serve()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2039 +0x13c4
  net/http.(*Server).Serve.gowrap3()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3285 +0x4f

Goroutine 118 (running) created at:
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness.go:231 +0xfcc
  k8s.io/apiserver/pkg/server/filters.(*priorityAndFairnessHandler).Handle-fm()
      <autogenerated>:1 +0x51
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2.1()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:192 +0xc1
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.func2()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:194 +0x312
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.newApfHandlerWithFilter.WithRequestInfo.func4()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/endpoints/filters/requestinfo.go:39 +0x17a
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  k8s.io/apiserver/pkg/server/filters.TestApfWatchPanic.func3()
      /home/akashem/go/src/k8s.io/kubernetes/staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:539 +0xc4
  net/http.HandlerFunc.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2166 +0x47
  net/http.serverHandler.ServeHTTP()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3137 +0x2a1
  net/http.(*conn).serve()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:2039 +0x13c4
  net/http.(*Server).Serve.gowrap3()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3285 +0x4f

Goroutine 114 (running) created at:
  net/http.(*Server).Serve()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/server.go:3285 +0x8ec
  net/http/httptest.(*Server).goServe.func1()
      /home/akashem/go/pkg/mod/golang.org/toolchain@v0.0.1-go1.22.0.linux-amd64/src/net/http/httptest/server.go:310 +0xbb
==================
    priority-and-fairness_test.go:189: Wanted 0 requests executing, got 1
    testing.go:1398: race detected during execution of test
--- FAIL: TestApfWatchPanic (0.01s)
FAIL
FAIL	k8s.io/apiserver/pkg/server/filters	0.103s
FAIL
```

When I run with the handler fix, I don't see the data race, and the test does not flake:
```
$ go test -race k8s.io/apiserver/pkg/server/filters -c

$ stress ./filters.test -test.run=TestApf
5s: 0 runs so far, 0 failures
10s: 8 runs so far, 0 failures
15s: 16 runs so far, 0 failures
20s: 24 runs so far, 0 failures
...
6m30s: 480 runs so far, 0 failures
6m35s: 488 runs so far, 0 failures
6m40s: 496 runs so far, 0 failures
```

tkashem commented:
(will do more `stress` test)
/hold

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/server/filters/priority-and-fairness_test.go:
```diff
@@ -177,11 +177,21 @@ func newApfHandlerWithFilter(t *testing.T, flowControlFilter utilflowcontrol.Int
-		if atomicReadOnlyExecuting != 0 {
-			t.Errorf("Wanted %d requests executing, got %d", 0, atomicReadOnlyExecuting)
-		}
+		func() {
```
This change is harmless. I still do not see whether/why it is necessary. That's because I still do not understand what is going wrong.

MikeSpreitzer commented:
/lgtm

MikeSpreitzer reviewed (approved):
/approve

MikeSpreitzer commented:
I created an extended version of this PR, please see #127313. I attempted to respond to the concerns about control flow from @neolit123 and @deads2k by (a) introducing and using `controlflow.TryFinally` to make the control flow clearer and (b) better documenting what is going on with the error relaying through that `resultCh`.

tkashem commented:
@MikeSpreitzer if both goroutines panic, unlikely but it's something I think is worth addressing? In that vein I have added two new commits
- https://github.com/kubernetes/kubernetes/pull/127029/commits/b28ff3522befd49df0e490f60627ef1ac5e21e62: the APF filters uses the abstraction that propagates panic value/stactrace accordingly
- https://github.com/kubernetes/kubernetes/pull/127029/commits/2df0f897b9de3641121b82577e732a836f9c4298 This package provides the abstraction to propagate panic from spawned goroutine to a parent goroutine

MikeSpreitzer commented:
@tkashem: I think that we have an even simpler fix, and part of the previous work is incomplete.

https://github.com/kubernetes/kubernetes/pull/127029/commits/a6b7435fd9d11f7772d103bc23198fa563fe0a67 makes sure that the bookkeeping updates _eventually_ happen, but does not guarantee that return from the handler "happens after" the bookkeeping updates. #127340 demonstrates this.

I think that we do not need to worry about relaying a panic-recovered value from the forked goroutine into the main goroutine because the forked goroutine only runs APF code and that is not supposed to panic ---- not even when timing out (I was mistaken earlier when I wrote that the use case of the relay is for timeout).

I have extended #127313 to add the necessary "happens after" linkage and remove the recovered value relaying; I also added a precondition check in the unit tests to guard against lingering problems from earlier tests.

MikeSpreitzer commented:
I have realized that my previous comment was buggy. Commit a6b7435fd9d11f7772d103bc23198fa563fe0a67 _does_ establish that the main handler's return "happens after" completion of the forked waiting goroutine. However, when in #127313 I removed the panic relay, I also removed that bit of synchronization; I restored the sync by using a `sync.WaitGroup`.

MikeSpreitzer commented:
The reason for the relaying of the panic is that the code in the forked waiting goroutine is not trivial, so it could panic and that should be handled reasonably. The most obvious reasonable handling is to pass it up the main stack so that it gets handled in the apiserver's regular way (see `WithPanicRecovery` in the same package). The relay loses the lower part of the stack, so that is why the relay is accompanied by a stack capture.

I have added a commit to #127313 that restores the relaying of a panic from the forked waiting goroutine, and also handles the possibility of both goroutines panicking. I think the code structure there should be a little easier to follow, and does not involve introducing a substantial supporting abstraction.

MikeSpreitzer commented:
@neolit123, @aojea, @deads2k: The reason for the relaying of the panic is that the code in the forked waiting goroutine is not trivial, so it could panic and that should be handled reasonably. The most obvious reasonable handling is to pass it up the main stack so that it gets handled in the apiserver's regular way (see `WithPanicRecovery` in the same package). The relay loses the lower part of the stack, so that is why the relay is accompanied by a stack capture.

I have added a commit to #127313 that restores the relaying of a panic from the forked waiting goroutine, and also handles the possibility of both goroutines panicking. I think the code structure there should be a little easier to follow, and does not involve introducing a substantial supporting abstraction.

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/panic/panic.go:
```diff
@@ -0,0 +1,198 @@
+		own := recover()
+		other := <-i.ch
+		if err := WrapRecoveredPanic(own, other); err != nil {
+			// both goroutines have panicked
```
s/both/one or both/

tkashem replied:
fixed in https://github.com/kubernetes/kubernetes/pull/127075

MikeSpreitzer commented on staging/src/k8s.io/apiserver/pkg/util/panic/panic.go:
```diff
@@ -0,0 +1,198 @@
+	// error object, so it can be propagated to the parent (request
+	// handler goroutine) via a channel
+	// The request handler uses this function to spawn a new goroutine
+	InvokeWithPanicProtection(func())
```
The names of the two "Invoke..." methods are so similar that I find it hard to remember which is which. For example, it is not obvious from the names that one creates a separate goroutine and the other does not.

tkashem replied:
renamed to `WrapPanic` in https://github.com/kubernetes/kubernetes/pull/127075

tkashem commented:
@MikeSpreitzer this PR reproduces the bookkeeping issue, and fixes it. 

I also have a follow up PR https://github.com/kubernetes/kubernetes/pull/127075 that we can use to manage panic (if both goroutines panic, we want a structured way to propagate the panic values and the stack traces up the chain)

MikeSpreitzer commented:
We have three related PRs to consider.

1. This one, which just fixes the concurrency bug in the main code and the test.
2. #127313, which contains the two commits from this PR and adds more that simplify and clarify the code structure by (a) introducing and using `controlflow.TryFinally` and (b) reorganizing the signaling from the waiting goroutine to the main goroutine.
3. #127075, a  introduces an abstraction for relaying panics and uses it in APF, with an eye towards eventually also using that abstraction for the other cases mentioned in #115565.

wojtek-t commented:
/lgtm
/approve
/retest

Let's follow-up after that merges.

tkashem commented:
new `stress` test results:
```
$ go test -race k8s.io/apiserver/pkg/server/filters -c
```

```
$ stress -p=512 ./filters.test -test.run=TestApfWatchPanic
5s: 250 runs so far, 0 failures
10s: 606 runs so far, 0 failures
15s: 995 runs so far, 0 failures
...
5m0s: 23340 runs so far, 0 failures
5m5s: 23743 runs so far, 0 failures
5m10s: 24136 runs so far, 0 failures
```

```
$ stress -p=64 ./filters.test -test.run="TestApf"
5s: 0 runs so far, 0 failures
10s: 64 runs so far, 0 failures
15s: 128 runs so far, 0 failures
...
10m15s: 6150 runs so far, 0 failures
10m20s: 6208 runs so far, 0 failures
10m25s: 6272 runs so far, 0 failures
```

```
$ stress ./filters.test -test.run=
5s: 0 runs so far, 0 failures
10s: 0 runs so far, 0 failures
15s: 0 runs so far, 0 failures
20s: 0 runs so far, 0 failures
25s: 0 runs so far, 0 failures
30s: 0 runs so far, 0 failures
35s: 0 runs so far, 0 failures
40s: 0 runs so far, 0 failures
45s: 6 runs so far, 0 failures
50s: 10 runs so far, 0 failures
...
10m10s: 150 runs so far, 0 failures
10m15s: 150 runs so far, 0 failures
10m20s: 151 runs so far, 0 failures
```

tkashem commented:
/hold cancel

(based on the new stress test result)
